Code/ForceField/ANI/AtomicContrib.cpp

```diff
@@ -0,0 +1,224 @@
+//
+//  Copyright (C) 2020 Manan Goel
+//
+//   @@ All Rights Reserved @@
+//  This file is part of the RDKit.
+//  The contents are covered by the terms of the BSD license
+//  which is included in the file license.txt, found at the root
+//  of the RDKit source tree.
+//
+
+#include "AtomicContrib.h"
+#include <ForceField/ForceField.h>
+#include <RDGeneral/Invariant.h>
+#include <RDGeneral/utils.h>
+#include <Numerics/EigenSerializer/EigenSerializer.h>
+#include <Eigen/Dense>
+#include <GraphMol/Descriptors/AtomicEnvironmentVector.h>
+#include <fstream>
+using namespace Eigen;
+
+namespace ForceFields {
+namespace ANI {
+ANIAtomContrib::ANIAtomContrib(ForceField *owner, int atomType,
+                               unsigned int atomIdx, VectorXi speciesVec,
+                               unsigned int numAtoms, unsigned int numLayers,
+                               unsigned int ensembleSize,
+                               std::string modelType) {
+  PRECONDITION(owner, "Bad Owner")
+  PRECONDITION(atomType == 0 || atomType == 1 || atomType == 2 || atomType == 3,
+               "Atom Type not Supported");
+  PRECONDITION(modelType == "ANI-1x" || modelType == "ANI-1ccx",
+               "Model Not currently supported")
+  PRECONDITION(ensembleSize > 0,
+               "There must be at least 1 model in the ensemble");
+  URANGE_CHECK(atomIdx, numAtoms);
+  dp_forceField = owner;
+  this->d_atomType = atomType;
+  this->d_atomIdx = atomIdx;
+  this->d_speciesVec = speciesVec;
+  this->d_numAtoms = numAtoms;
+  this->d_ensembleSize = ensembleSize;
+  this->d_modelType = modelType;
+
+  if (this->d_atomEncoding.find(this->d_atomType) !=
+      this->d_atomEncoding.end()) {
+    auto atomicSymbol = this->d_atomEncoding[this->d_atomType];
+    for (unsigned int modelNum = 0; modelNum < ensembleSize; modelNum++) {
+      std::vector<ArrayXXd> currModelWeights;
+      std::vector<ArrayXXd> currModelBiases;
+      Utils::loadFromBin(&currModelWeights, &currModelBiases, modelNum,
+                         atomicSymbol, this->d_modelType);
+      this->d_weights.push_back(currModelWeights);
+      this->d_biases.push_back(currModelBiases);
+    }
+    Utils::loadSelfEnergy(&(this->d_selfEnergy), atomicSymbol,
+                          this->d_modelType);
+  } else {
+    this->d_selfEnergy = 0;
+  }
+}
+
+double ANIAtomContrib::forwardProp(ArrayXXd aev) const {
+  if (this->d_atomType == -1) {
+    return 0;
+  }
+  if (aev.cols() != 1) {
+    aev.transposeInPlace();
+  }
+
+  std::vector<double> energies;
```

Comment:
since you know how big energies will be, you might as well go ahead and reserve that much space:
```suggestion
  std::vector<double> energies;
  energies.reserve(this->d_weights.size());
```